Modules/Portfolio/classes/class.ilPortfolioGlobalTemplate.php

```diff
@@ -0,0 +1,2059 @@
+	 * @param int $a_width banner width
+	 * @param int $a_height banner height
+	 * @param bool $a_export
+	 */
+	public function setBanner($a_img, $a_width = 1370, $a_height = 100, $a_export = false)
+	{		
+		if($a_img)
+		{
+			if(!$a_export)
+			{
+				$a_img = ILIAS_HTTP_PATH."/".$a_img;
+			}
+			
+			$this->setCurrentBlock("banner_bl");
+			$this->setVariable("BANNER_WIDTH", $a_width); // currently not needed
+			$this->setVariable("BANNER_HEIGHT", $a_height);
+			$this->setVariable("BANNER_URL", $a_img);			
+			$this->parseCurrentBlock();
+		}
+	}
+	
+	/**
+	 * Reset all header properties: title, icon, description, alerts, action menu
+	 */
+	public function resetHeaderBlock($a_reset_header_action = true)
+	{
+		$this->setTitle(null);
+		$this->setTitleIcon(null);
+		$this->setDescription(null);
+		$this->setAlertProperties(array());		
+		$this->enableDragDropFileUpload(null);
+		
+		// see setFullscreenHeader()
+		if($a_reset_header_action)
+		{
+			$this->setHeaderActionMenu(null);
+		}
+	}	
+	
+	/**
+	 * Enables the file upload into this object by dropping a file.
+	 */
+	public function enableDragDropFileUpload($a_ref_id)
+	{
+		$this->enable_fileupload = $a_ref_id;
+	}
+
+
+	// TEMPLATING AND GLOBAL RENDERING
+	//
+	// Forwards to ilTemplate-member.
+
+	/**
+	 * @param	string
+	 * @return	string
+	 */
+	public function get($part = "DEFAULT") {
+		return $this->template->get($part);
+	}
+
+	public function setVariable($variable, $value = '') {
+		return $this->template->setVariable($variable, $value);
+	}
+
+	private function variableExists($a_variablename)
+	{
+		return $this->template->variableExists($a_variablename);
+	}
+
+	/**
+	* @access	public
+	* @param	string
+	* @return	???
+	*/
+	public function setCurrentBlock ($part = "DEFAULT")
+	{
+		return $this->template->setCurrentBlock($part);
+	}
+
+	/**
+	* overwrites ITX::touchBlock.
+	* @access	public
+	* @param	string
+	* @return	???
+	*/
+	public function touchBlock($block)
+	{
+		return $this->template->touchBlock($block);
+	}
+
+	/**
+	* Überladene Funktion, die auf den aktuelle Block vorher noch ein replace ausführt
+	* @access	public
+	* @param	string
+	* @return	string
+	*/
+	public function parseCurrentBlock($part = "DEFAULT")
+	{
+		return $this->template->parseCurrentBlock($part);
+	}
+
+	/**
+	* overwrites ITX::addBlockFile
+	* @access	public
+	* @param	string
+	* @param	string
+	* @param	string		$tplname		template name
+	* @param	boolean		$in_module		should be set to true, if template file is in module subdirectory
+	* @return	boolean/string
+	*/
+	public function addBlockFile($var, $block, $tplname, $in_module = false)
+	{
+		return $this->template->addBlockFile($var, $block, $tplname, $in_module);
+	}
+
+	/**
+	* check if block exists in actual template
+	* @access	private
+	* @param string blockname
+	* @return	boolean
+	*/
+	public function blockExists($a_blockname)
+	{
+		return $this->template->blockExists($a_blockname);
+	}
+}
```

Comment:
Yes, it is. As explained in the description of the PR and in the answer to @mjansenDatabay, this is a tactical move to make the interface of the remaining `ilGlobalTemplate` a lot smaller, specifically get rid of methods that determine rendering order (fill*) and move the responsibility for that functionality closer to its users. I'm happy to supply my excel-sheet that I build to understand the usages of certain methods of ilTemplate. Any suggestions where I could put it?

Reply:
Maybe upload it to the FW page? Could these classes extend ilGlobalTemplate instead or are there any reasons to not do this?

Reply:
[This](https://docu.ilias.de/goto_docu_wiki_wpage_5251_1357.html#ilPageTocA114) is the afforementioned excel-sheet.

I think it might be possible that the class could extend `ilGlobalTemplate` but I am not 100% sure if this will really make things better or easier. I am sure that it would need to copy some methods (that are `private` now) to its own body. It might also be necessary to copy some none-private methods that refer to private methods in the original class.  It wouldn't cut it to just set the methods to `protected` in `ilGlobalTemplate` because then they would be part of the interface again (and would need to be considered in further refactoring).

Reply:
But most of these things will be moved out of this class in the future somehow if not completely? There will be some Screen KS elements, there will be the GlobalScreen service. The service which now intends to control main menu entries will most probably also control the footer, the slots of the screen (left, main, right) and so on. So my current expectation is that we will get rid of "new ilGlobalTemplate" code in the future and need some final interface in the GlobalScreen service anyway (and more KS elements supporting this).

But even in an interim solution I would like to get rid of these "clones", since it will lead to issues with almost every code change in ilGlobalTemplate that is not copied into the clones since they are highly coupled by using the same templates. I would need to have a detailed look into one of these cases. But I expect having the functions protected for now makes it easier and cleaner in these classes. Even if we introduce a dicto rule that says "do not extend ilGlobalTemplate" to make clear that this code needs to be refactored in the future again.

In general cross-component class inheritance is more a problem than a solution (yes, we did this for years).